### PR TITLE
Added haxe.Http support for nodejs

### PR DESCRIPTION
**Pull request by Blank101**

### CODE REVIEW

**Comment by ncannasse** [comment]
@nadako can you check + merge ?

**Comment by nadako** [comment]
I haven't used `haxe.Http` myself, but sure, I'll check it out :-)

**Comment by nadako** [comment]
We don't have tests for `haxe.Http`, right?

**Comment by Simn** [comment]
Not really, last time I tried to set that up it ended in tears.

**Comment by nadako** [comment]
@Blank101 It's missing `onStatus` functionality, also I think it makes sense to make `requestUrl` unavailable with `-D nodejs` since there's no way to do sync request on node.js and it'll always return null.

**Comment by Blank101** [comment]
Oops missed the onStatus, and yes requestUrl, withCredentials and async should be unavailable.

**Comment by nadako** [comment]
Another thing that's missing is setting multiple headers of the same name. It seems that it's also missing in `hxnodejs` externs though. We'll have to look at how to do it in node.js.

File: std/haxe/Http.hx [hunk 1]
```diff
@@ -196,7 +198,66 @@ class Http {
...
+			uri += StringTools.urlEncode(p.param)+"="+StringTools.urlEncode(p.value);
+		}
+		var question = path.split("?").length <= 1;
+		if (!post && uri != null) path += (if( question ) "?" else "&") + uri;
+		
+		var opts = {
+			protocol: secure ? "https:" : "http:",
+			host: host,
+			port: port,
+			method: post == true ? 'POST' : 'GET',
```
Comment (nadako) [inline std/haxe/Http.hx]: `post == true`, seriously? :)

**Comment by nadako** [comment]
added some more comments. to be honest, I think we should refactor `haxe.Http` asap

File: std/haxe/Http.hx (hunk 1 above)
Comment (Blank101) [inline std/haxe/Http.hx]: `post` could be null, no?

Comment (Blank101) [inline std/haxe/Http.hx]: lol I guess that doesn't make a difference for js target..

**Comment by Blank101** [comment]
Made the change, and yes I agree about refactoring haxe.Http. It's a mess right now.

**Comment by Simn** [comment]
I would also prefer to see this thing refactored before we add any more stuff to it...

File: std/haxe/Http.hx [hunk 2]
```diff
@@ -196,7 +198,66 @@ class Http {
 	**/
 	public function request( ?post : Bool ) : Void {
 		var me = this;
-	#if js
+	#if nodejs
+		me.responseData = null;
+		var url_regexp = ~/^(https?:\/\/)?([a-zA-Z\.0-9_-]+)(:[0-9]+)?(.*)$/;
```
Comment (nadako) [inline std/haxe/Http.hx]: Wouldn't it be easier to use `js.node.Url.parse` instead of own solution?

**Comment by nadako** [comment]
Okay, about multiple headers with same name, it seems that node.js [supports passing array of string instead of string](https://nodejs.org/api/http.html#http_response_setheader_name_value) for the header value. I'll update the hxnodejs extern, meanwhile please add support for that in `haxe.Http`.

File: std/haxe/Http.hx (hunk 2 above)
Comment (Blank101) [inline std/haxe/Http.hx]: Would probably make it easier to maintain. I just copied the regex from the customRequest.

**Comment by Blank101** [comment]
The resource you posted is for http response headers, not request.

**Comment by nadako** [comment]
Yes, but actually `getHeader`/`setHeader` are methods of `OutgoingMessage` class that is base for both `ClientRequest` and `ServerResponse`. Node.js documentation is really awful.

**Comment by nadako** [comment]
And those methods are used to set headers passed as `headers` field to the request options.

**Comment by Blank101** [comment]
Are we able to merge this in for the time being, and defer the refactor to some point in the future? This is a pretty critical feature for nodejs users.

**Comment by nadako** [comment]
I don't know, this introduces dependency on hxnodejs in the std code and I'm not sure this is a good idea. I'd rather go for https://github.com/HaxeFoundation/hxnodejs/issues/69

**Comment by nadako** [comment]
This can't go into hxnodejs because of https://github.com/HaxeFoundation/hxnodejs/issues/42, so I just merged this at last.

**Comment by back2dos** [comment]
> This can't go into hxnodejs because of HaxeFoundation/hxnodejs#42, so I just merged this at last.

Can you elaborate? I offered a solution for having a node-specific Bytes implementation but I fail to see how it's related to Http.

I really think we should either add nodejs to the std lib or put http for nodejs into hxnodejs. This kind of mix is pretty horrible.

**Comment by nadako** [comment]
> Can you elaborate?

If I just redefine haxe.Http in hxnodejs it will break usage of haxe.Http macros. I tried adding it to `js/_std/haxe/Http.hx` but that doesn't seem to work for libraries either. There's an open issue discussing a possible solution: #4249, but that's for the future.

I'm yet to try the solution you suggested here https://github.com/HaxeFoundation/hxnodejs/issues/42#issuecomment-233547042. It sounds like it might actually work, but still feels hacky to me (still an option tho).

> I really think we should either add nodejs to the std lib or put http for nodejs into hxnodejs. This kind of mix is pretty horrible.

Well we used to depend on hxssl for the https support, this isn't very different.

**Comment by back2dos** [comment]
> I'm yet to try the solution you suggested here HaxeFoundation/hxnodejs#42 (comment). It sounds like it might actually work, but still feels hacky to me (still an option tho).

I'm using this technique quite a lot. Maybe it's a hack (I guess overwriting a std class probably always will be). But it is isolated and allows for clean separation on the scale that really matters. Once better features are available, it is trivial to replace the hack.

> Well we used to depend on hxssl for the https support, this isn't very different.

Yes and that was horrible. I agree this isn't very different, particularly in that this is horrible too :D

**Pull request merged**